### PR TITLE
[tune] Ensure Cleanup

### LINKED ISSUE #7963: [tune] [ray] _stop in Trainable not running to completion
*Ray version and other system information (Python version, TensorFlow version, OS):*
ray 0.8.4
python 3.6.9

### What is the problem?
We're using _stop in our Trainable class to upload the weights files and call other end-of-training functions.

When uploading a file using boto3's S3Transfer or using `subprocess.Popen("aws s3 cp source target", shell=True)`, the code exits too soon and doesn't complete. 

Here's the rough flow:
_stop() -> our_function() -> upload_function() -> exit

So, as soon as the upload_function makes a call to a different subprocess, the code seems to exit and the remaining lines of code in _stop are left unexecuted.

Observations:
1. The code seems to exit exactly at the point where a different subprocess is called:
a. future.result() in boto3.s3.transfer.S3Transfer.upload_file.
b. After calling subprocess.Popen().
2. If local_mode=True is passed into ray.init, there are no problems and the code in _stop executes fully.
3. When a single trial is called, this error pops up every single time without fail.
4. For larger experiments with n trials, (n-1) trials have no issue, but for the final trial, the code exits before _stop completes.

Tested Fixes:
1. [crude] Calling `time.sleep` just after `tune.run` so that the _stop function has time to finish its execution.
2. In _stop_trial of RayTrialExecutor, get an object ID from the `trial.runner.stop.remote()` line and call `ray.get(object_id)`. This works perfectly.

```
object_id = trial.runner.stop.remote()
ray.get(object_id)
trial.runner.__ray_terminate__.remote()
```

Question:
While fix 2 above seems to work fine, there's no visibility for any outputs from _stop since it seems to run in the background and produces no logs. What could be a better way to resolve this?

Note:
Wasn't able to provide code to reproduce this issue without the use of external libraries like boto3 or the AWS CLI.

cc: @richardliaw

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Closes https://github.com/ray-project/ray/issues/7963.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/ray_trial_executor.py [hunk 1]
```diff
@@ -38,6 +39,41 @@ def unwrap(self):
...
+    Args:
+        threshold (int): Number of futures to hold at once. If the threshold
+            is passed, cleanup will kick in and remove futures.
+    """
+
+    def __init__(self, threshold=TRIAL_CLEANUP_THRESHOLD):
+        self.threshold = threshold
+        self._cleanup_map = {}
+
+    def add(self, trial, actor):
```
Comment (ujvl) [inline python/ray/tune/ray_trial_executor.py:54]: nit: docstrings for add and cleanup

File: python/ray/tune/ray_trial_executor.py [hunk 2]
```diff
@@ -38,6 +39,41 @@ def unwrap(self):
...
+        num_to_keep = int(self.threshold) / 2 if partial else 0
+        while len(self._cleanup_map) > num_to_keep:
+            dones, _ = ray.wait(
+                list(self._cleanup_map), timeout=DEFAULT_GET_TIMEOUT)
+            if not dones:
+                logger.warning(
+                    "Skipping cleanup - trainable.stop did not return in "
+                    "time. Consider making `stop` a faster operation.")
+            else:
+                done = dones[0]
```
Comment (ujvl) [inline python/ray/tune/ray_trial_executor.py:73]: wait why just delete the first one and not all

Comment (richardliaw) [reply python/ray/tune/ray_trial_executor.py:73]: `ray.wait` should return 1 entry at a time

File: python/ray/tune/ray_trial_executor.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/tune/ray_trial_executor.py:54]: done

**Review by ujvl — APPROVED** [review]

**Pull request merged**